File: Leetcode Daily Challenge/032-February-2026/008-balanced-binary-tree.py
# https://leetcode.com/problems/balanced-binary-tree/




# Definition for a binary tree node.
# class TreeNode:
def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
    def height(self, root):
        if not root:
            return 0

        L = self.height(root.left)
        if L == -1:
            return -1

        R = self.height(root.right)
        if R == -1:
            return -1

        if abs(L - R) > 1:
            return -1

        return 1 + max(L, R)

    # TC : O(N)
    # SC : O(H) Recursive Stack Space
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.height(root) != -1

File: Leetcode Daily Challenge/032-February-2026/008-balanced-binary-tree.cpp
// https://leetcode.com/problems/balanced-binary-tree/




#include <bits/stdc++.h>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    int height(TreeNode* root) {
        if (!root) return 0;

        int L = height(root -> left);
        if (L == -1) return -1;

        int R = height(root -> right);
        if (R == -1) return -1;

        if (abs(L - R) > 1) return -1;

        return 1 + max(L, R);
    }
public:
    bool isBalanced(TreeNode* root) {
        return height(root) != -1;
    }
};